File: apps/reading/src/api/workstationRecordPromptContextBridge.ts
/**
 * Workstation record pack → prompt context bridge (pure).
 *
 * Operator vision: records of insights/questions recursively inform all
 * prompts. This pure layer bridges a recursive record pack into a proposed
 * prompt envelope + optional model/budget decision snapshot.
 *
 * prompts_injected always false — never mutates live prompt pipelines.
 * record_persisted always false — never writes records.
 */

import {
  composeWorkstationRecursiveRecordPack,
  type WorkstationRecordItem,
  type WorkstationRecursiveRecordPack,
} from "./workstationRecursiveRecordPack";
import {
  composeModelDecisionWithProjection,
  type ModelDecisionPromptComposeResult,
  type ModelOption,
} from "./modelDecisionPromptCompose";

export type ContextPlacement = "prefix" | "suffix";

export interface WorkstationRecordPromptContextBridgeInput {
  session_id: string;
  /** User-facing prompt body (caller-supplied). */
  user_prompt: string;
  /** Record items to pack; required unless prebuilt_pack is supplied. */
  items?: WorkstationRecordItem[] | null;
  /**
   * Optional prebuilt pack (e.g. from prior compose). When set, items may be
   * omitted; pack must already be caller-supplied (not invented).
   */
  prebuilt_pack?: WorkstationRecursiveRecordPack | null;
  max_context_lines?: number | null;
  placement?: ContextPlacement;
  /** Optional model decision for budget projection of the enriched prompt. */
  model_decision?: {
    selected_model_id: string;
    models: ModelOption[];
    daily_cap_usd: number | null;
    spent_usd: number | null;
    projected_cost_usd_high?: number | null;
    projected_cost_usd_low?: number | null;
  } | null;
}

export interface PromptContextEnvelope {
  session_id: string;
  user_prompt: string;
  context_block: string;
  /** Full proposed prompt: context + user (order by placement). */
  proposed_prompt: string;
  context_line_count: number;
  placement: ContextPlacement;
  pack: WorkstationRecursiveRecordPack;
  /** Optional model/budget compose when requested. */
  model_decision: ModelDecisionPromptComposeResult | null;
  bridge_ready: boolean;
  /** Always false — pure bridge never injects into live prompts. */
  prompts_injected: false;
  /** Always false — pure bridge never persists records. */
  record_persisted: false;
  notes: string[];
  authority: "workstation_record_prompt_context_bridge_advisory";
}

function requireNonEmpty(value: unknown, name: string): string {
  if (typeof value !== "string" || !value.trim()) {
    throw new Error(`${name} must be a non-empty string`);
  }
  return value.trim();
}

function buildContextBlock(lines: string[]): string {
  if (lines.length === 0) return "";
  return (
    "### Workstation recursive context (advisory; caller-supplied only)\n" +
    lines.map((l) => `- ${l}`).join("\n")
  );
}

/**
 * Bridge workstation records into a proposed prompt envelope.
 * Never injects live; never invents record content.
 */
export function bridgeWorkstationRecordPromptContext(
  input: WorkstationRecordPromptContextBridgeInput,
): PromptContextEnvelope {
  if (!input || typeof input !== "object") {
    throw new Error("input must be an object");
  }
  const session_id = requireNonEmpty(input.session_id, "session_id");
  const user_prompt = requireNonEmpty(input.user_prompt, "user_prompt");

  const placement: ContextPlacement =
    input.placement === undefined || input.placement === null
      ? "prefix"
      : input.placement;
  if (placement !== "prefix" && placement !== "suffix") {
    throw new Error("placement must be prefix|suffix");
  }

  const notes: string[] = [
    "prompts_injected=false — proposed envelope only; no live injection",
    "record_persisted=false — bridge does not write records",
    "context lines are caller-supplied only (no invent)",
  ];

  let pack: WorkstationRecursiveRecordPack;
  if (input.prebuilt_pack != null) {
    const p = input.prebuilt_pack;
    if (!p || typeof p !== "object") {
      throw new Error("prebuilt_pack must be an object when set");
    }
    if (p.record_persisted !== false) {
      throw new Error("prebuilt_pack.record_persisted must be false");
    }
    if (p.prompts_injected !== false) {
      throw new Error("prebuilt_pack.prompts_injected must be false");
    }
    if (!Array.isArray(p.prompt_context_lines)) {
      throw new Error("prebuilt_pack.prompt_context_lines must be an array");
    }
    if (typeof p.session_id !== "string" || !p.session_id.trim()) {
      throw new Error("prebuilt_pack.session_id must be a non-empty string");
    }
    pack = p;
    notes.push("using prebuilt_pack (caller-supplied)");
  } else {
    if (!Array.isArray(input.items)) {
      throw new Error("items must be an array when prebuilt_pack is not set");
    }
    pack = composeWorkstationRecursiveRecordPack({
      session_id,
      items: input.items,
      max_context_lines: input.max_context_lines ?? null,
    });
    notes.push(...pack.notes);
  }

  const context_block = buildContextBlock(pack.prompt_context_lines);
  const proposed_prompt =
    context_block.length === 0
      ? user_prompt
      : placement === "prefix"
        ? `${context_block}\n\n### User prompt\n${user_prompt}`
        : `### User prompt\n${user_prompt}\n\n${context_block}`;

  if (context_block.length === 0) {
    notes.push(
      "context_block empty — proposed_prompt is user_prompt only (no invent context)",
    );
  } else {
    notes.push(
      `context_lines=${pack.prompt_context_lines.length} placement=${placement}`,
    );
  }

  let model_decision: ModelDecisionPromptComposeResult | null = null;
  if (input.model_decision != null) {
    if (typeof input.model_decision !== "object") {
      throw new Error("model_decision must be an object when set");
    }
    model_decision = composeModelDecisionWithProjection({
      selected_model_id: input.model_decision.selected_model_id,
      models: input.model_decision.models,
      daily_cap_usd: input.model_decision.daily_cap_usd,
      spent_usd: input.model_decision.spent_usd,
      projected_cost_usd_high: input.model_decision.projected_cost_usd_high,
      projected_cost_usd_low: input.model_decision.projected_cost_usd_low,
    });
    notes.push(...model_decision.notes);
    notes.push(
      `model_decision attached for selected=${model_decision.selected_model_id}`,
    );
  }

  const bridge_ready =
    user_prompt.length > 0 &&
    (pack.pack_ready || pack.prompt_context_lines.length === 0);
  // Bridge is ready with empty pack (user prompt alone) or with pack_ready.
  notes.push(
    bridge_ready
      ? "bridge_ready=true — proposed envelope prepared (not injected)"
      : "bridge_ready=false",
  );
  notes.push("prompts_injected=false");
  notes.push("record_persisted=false");

  return {
    session_id,
    user_prompt,
    context_block,
    proposed_prompt,
    context_line_count: pack.prompt_context_lines.length,
    placement,
    pack,
    model_decision,
    bridge_ready,
    prompts_injected: false,
    record_persisted: false,
    notes,
    authority: "workstation_record_prompt_context_bridge_advisory",
  };
}

export function formatWorkstationRecordPromptContextBridgeSummary(
  e: PromptContextEnvelope,
): string {
  return (
    `prompt bridge · session=${e.session_id} · lines=${e.context_line_count} · ` +
    `ready=${e.bridge_ready} · prompts_injected=false · record_persisted=false`
  );
}

File: substrate/workstation_record_prompt_context_bridge.py
"""Workstation record pack → prompt context bridge (pure).

Bridges recursive records into a proposed prompt envelope.
prompts_injected and record_persisted always False.
"""

from __future__ import annotations

from dataclasses import dataclass
from typing import Any, Literal

from substrate.model_decision.prompt_compose import (
    ModelDecisionPromptComposeError,
    ModelDecisionPromptComposeResult,
    compose_model_decision_with_projection,
)
from substrate.workstation_recursive_record_pack import (
    WorkstationRecursiveRecordPack,
    WorkstationRecursiveRecordPackError,
    compose_workstation_recursive_record_pack,
)

ContextPlacement = Literal["prefix", "suffix"]


class WorkstationRecordPromptContextBridgeError(ValueError):
    """Fail-closed validation for record→prompt context bridge."""


@dataclass(frozen=True)
class PromptContextEnvelope:
    session_id: str
    user_prompt: str
    context_block: str
    proposed_prompt: str
    context_line_count: int
    placement: str
    pack: WorkstationRecursiveRecordPack
    model_decision: ModelDecisionPromptComposeResult | None
    bridge_ready: bool
    prompts_injected: bool
    record_persisted: bool
    notes: tuple[str, ...]
    authority: str

    def to_dict(self) -> dict[str, Any]:
        return {
            "session_id": self.session_id,
            "user_prompt": self.user_prompt,
            "context_block": self.context_block,
            "proposed_prompt": self.proposed_prompt,
            "context_line_count": self.context_line_count,
            "placement": self.placement,
            "pack": self.pack.to_dict(),
            "model_decision": (
                self.model_decision.to_dict()
                if self.model_decision is not None
                else None
            ),
            "bridge_ready": self.bridge_ready,
            "prompts_injected": False,
            "record_persisted": False,
            "notes": list(self.notes),
            "authority": "workstation_record_prompt_context_bridge_advisory",
        }


def _require_nonempty(value: object, *, field: str) -> str:
    if not isinstance(value, str) or not value.strip():
        raise WorkstationRecordPromptContextBridgeError(
            f"{field} must be a non-empty string"
        )
    return value.strip()


def _build_context_block(lines: list[str] | tuple[str, ...]) -> str:
    if not lines:
        return ""
    body = "\n".join(f"- {l}" for l in lines)
    return (
        "### Workstation recursive context (advisory; caller-supplied only)\n"
        + body
    )


def bridge_workstation_record_prompt_context(
    *,
    session_id: object,
    user_prompt: object,
    items: object | None = None,
    prebuilt_pack: object | None = None,
    max_context_lines: object | None = None,
    placement: object | None = None,
    model_decision: object | None = None,
) -> PromptContextEnvelope:
    """Bridge records into proposed prompt envelope. Never injects live."""
    sid = _require_nonempty(session_id, field="session_id")
    prompt = _require_nonempty(user_prompt, field="user_prompt")

    place: str = "prefix"
    if placement is not None:
        if placement not in ("prefix", "suffix"):
            raise WorkstationRecordPromptContextBridgeError(
                "placement must be prefix|suffix"
            )
        place = placement  # type: ignore[assignment]

    notes: list[str] = [
        "prompts_injected=false — proposed envelope only; no live injection",
        "record_persisted=false — bridge does not write records",
        "context lines are caller-supplied only (no invent)",
    ]

    pack: WorkstationRecursiveRecordPack
    if prebuilt_pack is not None:
        if not isinstance(prebuilt_pack, WorkstationRecursiveRecordPack):
            # Allow dict-shaped pack for route layer convenience
            if not isinstance(prebuilt_pack, dict):
                raise WorkstationRecordPromptContextBridgeError(
                    "prebuilt_pack must be an object when set"
                )
            if prebuilt_pack.get("record_persisted") is not False:
                raise WorkstationRecordPromptContextBridgeError(
                    "prebuilt_pack.record_persisted must be false"
                )
            if prebuilt_pack.get("prompts_injected") is not False:
                raise WorkstationRecordPromptContextBridgeError(
                    "prebuilt_pack.prompts_injected must be false"
                )
            lines = prebuilt_pack.get("prompt_context_lines")
            if not isinstance(lines, list):
                raise WorkstationRecordPromptContextBridgeError(
                    "prebuilt_pack.prompt_context_lines must be an array"
                )
            pack_sid = _require_nonempty(
                prebuilt_pack.get("session_id"),
                field="prebuilt_pack.session_id",
            )
            # Reconstruct via compose from empty items not allowed — use
            # lightweight frozen rebuild for dict prebuilt only when lines ok
            pack = WorkstationRecursiveRecordPack(
                session_id=pack_sid,
                item_count=int(prebuilt_pack.get("item_count") or len(lines)),
                by_kind=dict(
                    prebuilt_pack.get("by_kind")
                    or {
                        "insight": 0,
                        "question": 0,
                        "highlight": 0,
                        "finding": 0,
                        "open_thread": 0,
                    }
                ),
                prompt_context_lines=tuple(str(x) for x in lines),
                pack_ready=bool(prebuilt_pack.get("pack_ready")),
                record_persisted=False,
                prompts_injected=False,
                notes=tuple(prebuilt_pack.get("notes") or ()),
                authority="workstation_recursive_record_pack_advisory",
            )
        else:
            if prebuilt_pack.record_persisted is not False:
                raise WorkstationRecordPromptContextBridgeError(
                    "prebuilt_pack.record_persisted must be false"
                )
            if prebuilt_pack.prompts_injected is not False:
                raise WorkstationRecordPromptContextBridgeError(
                    "prebuilt_pack.prompts_injected must be false"
                )
            pack = prebuilt_pack
        notes.append("using prebuilt_pack (caller-supplied)")
    else:
        if not isinstance(items, list):
            raise WorkstationRecordPromptContextBridgeError(
                "items must be an array when prebuilt_pack is not set"
            )
        try:
            pack = compose_workstation_recursive_record_pack(
                session_id=sid,
                items=items,
                max_context_lines=max_context_lines,
            )
        except WorkstationRecursiveRecordPackError as e:
            raise WorkstationRecordPromptContextBridgeError(str(e)) from e
        notes.extend(pack.notes)

    context_block = _build_context_block(pack.prompt_context_lines)
    if not context_block:
        proposed = prompt
        notes.append(
            "context_block empty — proposed_prompt is user_prompt only (no invent context)"
        )
    elif place == "prefix":
        proposed = (
            f"{context_block}\n\n### User prompt\n{prompt}"
        )
        notes.append(
            f"context_lines={len(pack.prompt_context_lines)} placement=prefix"
        )
    else:
        proposed = (
            f"### User prompt\n{prompt}\n\n{context_block}"
        )
        notes.append(
            f"context_lines={len(pack.prompt_context_lines)} placement=suffix"
        )

    md: ModelDecisionPromptComposeResult | None = None
    if model_decision is not None:
        if not isinstance(model_decision, dict):
            raise WorkstationRecordPromptContextBridgeError(
                "model_decision must be an object when set"
            )
        try:
            md = compose_model_decision_with_projection(
                selected_model_id=model_decision.get("selected_model_id"),
                models=model_decision.get("models"),
                daily_cap_usd=model_decision.get("daily_cap_usd"),
                spent_usd=model_decision.get("spent_usd"),
                projected_cost_usd_high=model_decision.get(
                    "projected_cost_usd_high"
                ),
                projected_cost_usd_low=model_decision.get(
                    "projected_cost_usd_low"
                ),
            )
        except ModelDecisionPromptComposeError as e:
            raise WorkstationRecordPromptContextBridgeError(str(e)) from e
        notes.extend(md.notes)
        notes.append(
            f"model_decision attached for selected={md.selected_model_id}"
        )

    bridge_ready = True
    notes.append(
        "bridge_ready=true — proposed envelope prepared (not injected)"
    )
    notes.append("prompts_injected=false")
    notes.append("record_persisted=false")

    return PromptContextEnvelope(
        session_id=sid,
        user_prompt=prompt,
        context_block=context_block,
        proposed_prompt=proposed,
        context_line_count=len(pack.prompt_context_lines),
        placement=place,
        pack=pack,
        model_decision=md,
        bridge_ready=bridge_ready,
        prompts_injected=False,
        record_persisted=False,
        notes=tuple(notes),
        authority="workstation_record_prompt_context_bridge_advisory",
    )


__all__ = [
    "PromptContextEnvelope",
    "WorkstationRecordPromptContextBridgeError",
    "bridge_workstation_record_prompt_context",
]
